### ISSUE TITLE
perf(augmentation): RandomRain draws drops with per-image and per-row Python loops (204 adds + 100 index_put_ for a batch of 8); vectorise the drop rasterisation

### ISSUE THREAD

**Issue by ducha-aiki** [kornia/kornia#4530]
## Problem

`RandomRain.apply_transform` (`kornia/augmentation/_2d/intensity/random_rain.py:89–124`) loops over the batch and, inside, over every row of the drop line:

```python
for i in range(image.shape[0]):
    ...
    for k in range(x.shape[0]):
        modeified_img[i, :, coords[0] + x[k], coords[1] + y[k]] = 200 / 255
```

That is `B × size_of_line` indexed writes, each a separate kernel (plus `int(params[...][i])` syncs). Measured with the defaults used in the sweep (drop height 5–20, 100–200 drops):

| shape | CPU | MPS |
|---|---:|---:|
| 1×256² | 0.20 ms | 3.2 ms |
| 8×256² | 1.5 ms | **16.8 ms** |
| 32×128² | 4.4 ms | 47 ms |

One invocation at batch 8 issues 204 `add`, 100 `index_put_`, 16 `linspace`, 8 `cat`; the batched heavy augmentations (`RandomElasticTransform`, `RandomMotionBlur`, `RandomGaussianNoise`) are 4–12x *faster* on MPS at the same shape, so this is the implementation, not the device.

## Fix

Build all drop pixel coordinates at once: `coords[:, None, :] + offsets[None, :, :]` for the `(B, N_drops, size_of_line)` grid, clamp/mask to the image, and do a single `index_put_` per batch (or one overall, with the batch index as the first index tensor). The recent fixes #4451 and #4453 touched the same function and keep the drop-count semantics that the vectorised version must preserve (`same_on_batch` draws the same number of drops per sample).

Measured on torch nightly 2.15.0.dev20260913, Apple M1 8 GB, macOS 26.5.1, kornia `633951bd7`, `PYTORCH_ENABLE_MPS_FALLBACK` unset, `torch.utils.benchmark` medians with `mps.synchronize` inside the timed region, fresh process per op. Split out of the MPS-audit tracking issue #4521; the PyTorch-side kernel findings are in #4522.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).
